tests/proxy_overload_recovery.erl

```diff
@@ -0,0 +1,442 @@
+%%     true = is_process_alive(PPid),
+%%     true = is_process_alive(VPid),
+%%     true.
+
+
+%%
+%% Prepare runtime state, pick a vnode and find pids - precondition_common
+%% ensures this runs first.
+%%
+prepare_args(_S) ->
+    [threshold_seed()].
+
+prepare(ThresholdSeed) ->
+    {RequestInterval, Interval, Threshold} = get_params(ThresholdSeed),
+    Index = 0, % Mess with the same vnode proxy each time - could rotate if we wanted
+
+    %% Set (or unset) appenvs for the test
+    prep_env(vnode_check_request_interval, RequestInterval),
+    prep_env(vnode_check_interval, Interval),
+    prep_env(vnode_overload_threshold, Threshold),
+    prep_env(inactivity_timeout, infinity),
+
+    %% Reset the vnode and the proxy to get back into a known state.
+    %% Ask the supervisor to do it for us to avoid messy messages in the logs.
+    Id = 0,
+    {ok, VPid0} = riak_core_vnode_manager:get_vnode_pid(Id, riak_kv_vnode),
+    sys:resume(VPid0),
+    ok = supervisor:terminate_child(riak_core_vnode_sup, VPid0),
+    false = is_process_alive(VPid0),
+
+    %% Reset the proxy pid to make sure it resets state and picks up the new
+    %% environment variables
+    ok = supervisor:terminate_child(riak_core_vnode_proxy_sup, {riak_kv_vnode, Id}),
+    RegName = riak_core_vnode_proxy:reg_name(riak_kv_vnode, Index),
+    undefined = whereis(RegName),
+    {ok, PPid} = supervisor:restart_child(riak_core_vnode_proxy_sup, {riak_kv_vnode, Id}),
+
+    %% Find the proxy pid and check it's alive and matches the supervisor
+    RegName = riak_core_vnode_proxy:reg_name(riak_kv_vnode, Index),
+    PPid = whereis(RegName),
+
+    %% Send a message through the vnode proxy (the timeout message is ignored)
+    %% and return the Pid so we know we have the same Pid.
+    {ok, VPid} = riak_core_vnode_proxy:command_return_vnode(
+                   {riak_kv_vnode,Index,node()}, timeout),
+
+    true = is_process_alive(PPid),
+    true = is_process_alive(VPid),
+    #rt{id = Id, pl = {Id, node()}, ppid = PPid, vpid = VPid}.
+
+prepare_next(S, RT, [ThresholdSeed]) ->
+    {_RequestInterval, _Interval, Threshold} = get_params(ThresholdSeed),
+    S#tstate{rt = RT, threshold = Threshold}.
+
+
+%%
+%% Suspend the vnode so it cannot process messages and builds
+%% up the vnode message queue
+%%
+suspend_pre(#tstate{vnode_running = Running}) ->
+    Running.
+
+suspend_args(#tstate{rt = RT}) ->
+    [RT].
+
+suspend(#rt{vpid = VPid}) ->
+    sys:suspend(VPid).
+
+suspend_next(S, _V, _A) ->
+    S#tstate{vnode_running = false}.
+
+%%
+%% Resume the vnode and wait until it has completed
+%% processing all queued messages
+%%
+resume_pre(#tstate{vnode_running = Running}) ->
+    not Running.
+
+resume_args(#tstate{rt = RT}) ->
+    [RT].
+
+resume(#rt{ppid = PPid, vpid = VPid}) ->
+    sys:resume(VPid),
+    ok = drain(VPid),
+    %% make sure all messages have been processed and there
+    %% is no pending 'ping' in the vnode before we continue.
+    sys:get_status(PPid),
```

Comment:
Yes, it's a trick to make sure a synchronous call goes through the proxy pid without changing any state, guaranteeing any messages that have been sent will be processed before the next commands.  I'll clarify the comment further.